src/dict.c

```diff
@@ -286,18 +286,35 @@ int _dictExpand(dict *d, unsigned long size, int* malloc_failed)
     return DICT_OK;
 }
 
+int _dictExpand(dict *d, unsigned long size, int* malloc_failed) {
+    /* the size is invalid if it is smaller than the size of the hash table 
+     * or smaller than the number of elements already inside the hash table */
+    if (dictIsRehashing(d) || d->ht_used[0] > size || DICTHT_SIZE(d->ht_size_exp[0]) >= size)
+        return DICT_ERR;
+    return _dictResize(d, size, malloc_failed);
+}
+
 /* return DICT_ERR if expand was not performed */
 int dictExpand(dict *d, unsigned long size) {
```

Comment:
In theory, we should also check `dict_can_resize` just like what `dictShrinkToFit` does. However, there are certain situations where a forced expand is necessary, such as when `listpack` converts to `dict`. It might be appropriate to add a `dictExpandForce` interface in such cases.

Reply:
i think `dict_can_resize` should be used in the "automatic" mechanisms, like `dictShrinkToFit`, `_dictExpandIfNeeded`, and `_dictShrinkIfNeeded`. i think it's fine that we don't use it in the explicit `dictExpand`, and `dictShrink`.